## Практическая работа 3 - Себежко А.А. П50-1-22

Цель работы: Провести разведочный анализ данных по датасету. Изучить как группировать, фильтровать, сортировать, обрабатывать и делать срезы данных.

Ход работы:

Импортируем библиотеки `pandas` и `numpy` для исследования датасета.

In [176]:
import pandas as pd
import numpy as np

Выберем датасет поездок метро в Англии. Загрузим датасет и просмотрим информацию о нем.

In [177]:
df = pd.read_csv("railway.csv")
df

,Transaction ID,Date of Purchase,Time of Purchase,Purchase Type,Payment Method,Railcard,Ticket Class,Ticket Type,Price,Departure Station,Arrival Destination,Date of Journey,Departure Time,Arrival Time,Actual Arrival Time,Journey Status,Reason for Delay,Refund Request
0,da8a6ba8-b3dc-4677-b176,2023-12-08,12:41:11,Online,Contactless,Adult,Standard,Advance,43,London Paddington,Liverpool Lime Street,2024-01-01,11:00:00,13:30:00,13:30:00,On Time,NaN,No
1,b0cdd1b0-f214-4197-be53,2023-12-16,11:23:01,Station,Credit Card,Adult,Standard,Advance,23,London Kings Cross,York,2024-01-01,09:45:00,11:35:00,11:40:00,Delayed,Signal Failure,No
2,f3ba7a96-f713-40d9-9629,2023-12-19,19:51:27,Online,Credit Card,NaN,Standard,Advance,3,Liverpool Lime Street,Manchester Piccadilly,2024-01-02,18:15:00,18:45:00,18:45:00,On Time,NaN,No
3,b2471f11-4fe7-4c87-8ab4,2023-12-20,23:00:36,Station,Credit Card,NaN,Standard,Advance,13,London Paddington,Reading,2024-01-01,21:30:00,22:30:00,22:30:00,On Time,NaN,No
4,2be00b45-0762-485e-a7a3,2023-12-27,18:22:56,Online,Contactless,NaN,Standard,Advance,76,Liverpool Lime Street,London Euston,2024-01-01,16:45:00,19:00:00,19:00:00,On Time,NaN,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
31648,1304623d-b8b7-4999-8e9c,2024-04-30,18:42:58,Online,Credit Card,NaN,Standard,Off-Peak,4,Manchester Piccadilly,Liverpool Lime Street,2024-04-30,20:00:00,20:30:00,20:30:00,On Time,NaN,No
31649,7da22246-f480-417c-bc2f,2024-04-30,18:46:10,Online,Contactless,NaN,Standard,Off-Peak,10,London Euston,Birmingham New Street,2024-04-30,20:15:00,21:35:00,21:35:00,On Time,NaN,No
31650,add9debf-46c1-4c75-b52d,2024-04-30,18:56:41,Station,Credit Card,NaN,Standard,Off-Peak,4,Manchester Piccadilly,Liverpool Lime Street,2024-04-30,20:15:00,20:45:00,20:45:00,On Time,NaN,No
31651,b92b047c-21fd-4859-966a,2024-04-30,19:51:47,Station,Credit Card,NaN,Standard,Off-Peak,10,London Euston,Birmingham New Street,2024-04-30,21:15:00,22:35:00,22:35:00,On Time,NaN,No


Получим сводку информации по датафрейму. Из нее можно узнать что в этом датафрейме 18 столбцов и 31.653 строчки. А также название каждого столбца, размер датасета (4.3+ мб), тип данных и количество записей по каждому столбику.

In [178]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction ID       31653 non-null  object
 1   Date of Purchase     31653 non-null  object
 2   Time of Purchase     31653 non-null  object
 3   Purchase Type        31653 non-null  object
 4   Payment Method       31653 non-null  object
 5   Railcard             10735 non-null  object
 6   Ticket Class         31653 non-null  object
 7   Ticket Type          31653 non-null  object
 8   Price                31653 non-null  int64 
 9   Departure Station    31653 non-null  object
 10  Arrival Destination  31653 non-null  object
 11  Date of Journey      31653 non-null  object
 12  Departure Time       31653 non-null  object
 13  Arrival Time         31653 non-null  object
 14  Actual Arrival Time  29773 non-null  object
 15  Journey Status       31653 non-null  object
 16  Reas

Выведем % по столбцам которые равны `null`. Столбцы `Railcard` и `Reason for Delay`, а также `Actual Arrival Time` имеют пустые значения.

In [179]:
df.isnull().mean()*100

Transaction ID          0.000000
Date of Purchase        0.000000
Time of Purchase        0.000000
Purchase Type           0.000000
Payment Method          0.000000
Railcard               66.085363
Ticket Class            0.000000
Ticket Type             0.000000
Price                   0.000000
Departure Station       0.000000
Arrival Destination     0.000000
Date of Journey         0.000000
Departure Time          0.000000
Arrival Time            0.000000
Actual Arrival Time     5.939405
Journey Status          0.000000
Reason for Delay       86.819575
Refund Request          0.000000
dtype: float64

Очищать эти столбцы не имеет смысла, так как отсутсвия этих данных тоже данные. Соответсвенно их стоит заменить на более понятные текстовые данные, кроме столбца `Actual Arrival Time`, так как мы не знаем во сколько пришел поезд, то и остальные данные бессполезны, следовательно удалим `NaN`.

In [180]:
df = df.dropna(subset=["Actual Arrival Time"]).copy()

df['Railcard'] = df['Railcard'].fillna("Not used")
df['Reason for Delay'] = df['Reason for Delay'].fillna("No delay")

df.isnull().mean() * 100

Transaction ID         0.0
Date of Purchase       0.0
Time of Purchase       0.0
Purchase Type          0.0
Payment Method         0.0
Railcard               0.0
Ticket Class           0.0
Ticket Type            0.0
Price                  0.0
Departure Station      0.0
Arrival Destination    0.0
Date of Journey        0.0
Departure Time         0.0
Arrival Time           0.0
Actual Arrival Time    0.0
Journey Status         0.0
Reason for Delay       0.0
Refund Request         0.0
dtype: float64

Проверим сколько осталось строчек после удаления `NaN` значений.

In [181]:
df.index.size

29773

Удалим все дубликаты, если они имеются. После удаления, количество строчек не поменялось, следовательно дубликатов не было.

In [182]:
df = df.drop_duplicates()
df.shape[0]

29773

Выведем первые 5 значений.

In [183]:
df.head(5)

,Transaction ID,Date of Purchase,Time of Purchase,Purchase Type,Payment Method,Railcard,Ticket Class,Ticket Type,Price,Departure Station,Arrival Destination,Date of Journey,Departure Time,Arrival Time,Actual Arrival Time,Journey Status,Reason for Delay,Refund Request
0,da8a6ba8-b3dc-4677-b176,2023-12-08,12:41:11,Online,Contactless,Adult,Standard,Advance,43,London Paddington,Liverpool Lime Street,2024-01-01,11:00:00,13:30:00,13:30:00,On Time,No delay,No
1,b0cdd1b0-f214-4197-be53,2023-12-16,11:23:01,Station,Credit Card,Adult,Standard,Advance,23,London Kings Cross,York,2024-01-01,09:45:00,11:35:00,11:40:00,Delayed,Signal Failure,No
2,f3ba7a96-f713-40d9-9629,2023-12-19,19:51:27,Online,Credit Card,Not used,Standard,Advance,3,Liverpool Lime Street,Manchester Piccadilly,2024-01-02,18:15:00,18:45:00,18:45:00,On Time,No delay,No
3,b2471f11-4fe7-4c87-8ab4,2023-12-20,23:00:36,Station,Credit Card,Not used,Standard,Advance,13,London Paddington,Reading,2024-01-01,21:30:00,22:30:00,22:30:00,On Time,No delay,No
4,2be00b45-0762-485e-a7a3,2023-12-27,18:22:56,Online,Contactless,Not used,Standard,Advance,76,Liverpool Lime Street,London Euston,2024-01-01,16:45:00,19:00:00,19:00:00,On Time,No delay,No


Выведем последние 5 значений.

In [184]:
df.tail(5)

,Transaction ID,Date of Purchase,Time of Purchase,Purchase Type,Payment Method,Railcard,Ticket Class,Ticket Type,Price,Departure Station,Arrival Destination,Date of Journey,Departure Time,Arrival Time,Actual Arrival Time,Journey Status,Reason for Delay,Refund Request
31648,1304623d-b8b7-4999-8e9c,2024-04-30,18:42:58,Online,Credit Card,Not used,Standard,Off-Peak,4,Manchester Piccadilly,Liverpool Lime Street,2024-04-30,20:00:00,20:30:00,20:30:00,On Time,No delay,No
31649,7da22246-f480-417c-bc2f,2024-04-30,18:46:10,Online,Contactless,Not used,Standard,Off-Peak,10,London Euston,Birmingham New Street,2024-04-30,20:15:00,21:35:00,21:35:00,On Time,No delay,No
31650,add9debf-46c1-4c75-b52d,2024-04-30,18:56:41,Station,Credit Card,Not used,Standard,Off-Peak,4,Manchester Piccadilly,Liverpool Lime Street,2024-04-30,20:15:00,20:45:00,20:45:00,On Time,No delay,No
31651,b92b047c-21fd-4859-966a,2024-04-30,19:51:47,Station,Credit Card,Not used,Standard,Off-Peak,10,London Euston,Birmingham New Street,2024-04-30,21:15:00,22:35:00,22:35:00,On Time,No delay,No
31652,1d5d89a2-bde5-410f-8f91,2024-04-30,20:05:39,Station,Credit Card,Adult,Standard,Off-Peak,3,Liverpool Lime Street,Manchester Piccadilly,2024-04-30,21:30:00,22:00:00,22:00:00,On Time,No delay,No


Произведем срез по трем столбцам.

In [185]:
df[['Refund Request', 'Purchase Type', 'Departure Station']]

,Refund Request,Purchase Type,Departure Station
0,No,Online,London Paddington
1,No,Station,London Kings Cross
2,No,Online,Liverpool Lime Street
3,No,Station,London Paddington
4,No,Online,Liverpool Lime Street
...,...,...,...
31648,No,Online,Manchester Piccadilly
31649,No,Online,London Euston
31650,No,Station,Manchester Piccadilly
31651,No,Station,London Euston


Произведем срез с 800 по 1000 строчки начиная от столбца `Purchase Type` до `Departure Station`.

In [186]:
df.loc[800:1000, 'Purchase Type': 'Departure Station']

,Purchase Type,Payment Method,Railcard,Ticket Class,Ticket Type,Price,Departure Station
800,Online,Credit Card,Disabled,Standard,Off-Peak,7,London Euston
801,Station,Credit Card,Not used,Standard,Off-Peak,19,Reading
802,Online,Credit Card,Disabled,Standard,Off-Peak,7,London Euston
803,Station,Credit Card,Adult,Standard,Anytime,9,London Euston
804,Station,Credit Card,Adult,Standard,Anytime,9,London Euston
...,...,...,...,...,...,...,...
996,Online,Debit Card,Not used,Standard,Advance,72,London Euston
997,Online,Credit Card,Disabled,Standard,Advance,4,York
998,Online,Credit Card,Not used,First Class,Advance,52,London Euston
999,Online,Credit Card,Not used,Standard,Advance,18,London Paddington


Произведем срез с 100-й строки по 1000-ю начиная от 4 столбца по 7 столбец.

In [187]:
df.iloc[100:1000, 4:7]

,Payment Method,Railcard,Ticket Class
103,Contactless,Disabled,Standard
104,Contactless,Not used,Standard
105,Credit Card,Not used,Standard
106,Contactless,Not used,Standard
107,Credit Card,Not used,Standard
...,...,...,...
1068,Debit Card,Disabled,Standard
1069,Contactless,Adult,Standard
1070,Contactless,Senior,Standard
1071,Contactless,Senior,Standard


Поставим фильтр по онлайн покупкам, а потом отсортируем по цене билета по убыванию.

In [188]:
df_filtered = df[df['Purchase Type'] == 'Online']
df_sorted = df_filtered.sort_values('Price', ascending=False)
df_sorted

,Transaction ID,Date of Purchase,Time of Purchase,Purchase Type,Payment Method,Railcard,Ticket Class,Ticket Type,Price,Departure Station,Arrival Destination,Date of Journey,Departure Time,Arrival Time,Actual Arrival Time,Journey Status,Reason for Delay,Refund Request
14748,3f631fa5-8a15-4918-8b31,2024-02-19,05:12:29,Online,Debit Card,Not used,First Class,Anytime,216,London Euston,Manchester Piccadilly,2024-02-19,06:30:00,08:20:00,08:20:00,On Time,No delay,No
15645,df9d431f-02c5-4c85-bb6d,2024-02-26,05:00:07,Online,Debit Card,Not used,First Class,Anytime,216,London Euston,Manchester Piccadilly,2024-02-26,06:30:00,08:20:00,08:20:00,On Time,No delay,No
29564,b8f91da1-380b-42c5-81b7,2024-04-22,17:21:36,Online,Credit Card,Not used,First Class,Anytime,216,London Euston,Manchester Piccadilly,2024-04-22,18:45:00,20:35:00,20:35:00,On Time,No delay,No
17698,16c9b1c4-1c53-4e61-ae8b,2024-03-08,17:24:48,Online,Credit Card,Not used,First Class,Anytime,216,London Euston,Manchester Piccadilly,2024-03-08,18:45:00,20:35:00,20:35:00,On Time,No delay,No
949,9c1e0d3f-caed-4775-8bdc,2024-01-05,05:01:20,Online,Debit Card,Not used,First Class,Anytime,216,London Euston,Manchester Piccadilly,2024-01-05,06:30:00,08:20:00,08:20:00,On Time,No delay,No
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2488,39195f02-ae74-4902-bffd,2024-01-10,20:29:46,Online,Credit Card,Not used,Standard,Advance,1,Birmingham New Street,Wolverhampton,2024-01-11,18:45:00,19:00:00,19:00:00,On Time,No delay,No
16099,9ef8e91d-7ba1-4d72-b338,2024-03-02,20:18:31,Online,Credit Card,Not used,Standard,Advance,1,Birmingham New Street,Wolverhampton,2024-03-03,18:45:00,19:00:00,19:00:00,On Time,No delay,No
3855,e633318f-b30d-4e0a-9cbd,2024-01-15,20:19:15,Online,Credit Card,Not used,Standard,Advance,1,Birmingham New Street,Wolverhampton,2024-01-16,18:45:00,19:00:00,19:00:00,On Time,No delay,No
24481,93f73d08-d54f-4a29-9af0,2024-04-03,20:21:06,Online,Credit Card,Not used,Standard,Advance,1,Birmingham New Street,Wolverhampton,2024-04-04,18:45:00,19:00:00,19:00:00,On Time,No delay,No


Выведем сводную статистику всех числовых столбцов, такую как:

* количество всех билетов - 29773+
* среднее значение - 23
* стандартное отклонение - 29
* диапазон
* минимум - 1 за билет (вероятно, это какая-то льготная цена)
* максимум - 267 возможно, это ошибка или цена на какой-то специфический тип билета

квартили:

* 25% - 5 то есть 25% билетов стоят 5 или меньше.
* медиана - 11 это означает, что половина билетов стоит меньше 11, а половина — больше. Медиана значительно ниже среднего, что ещё раз подтверждает наличие высоких цен, влияющих на среднее значение.
* 75% - 35 то есть 75% билетов стоят 35 или меньше.



Анализ данных по 29773 билетам показывает, что средняя цена составляет 23.39. Однако, значительное стандартное отклонение (29.95) указывает на большой разброс цен. Это говорит о том, что цены на билеты сильно варьируются.



Этот метод позволяет быстро оценить масштаб, отклонение и диапазон числовых данных.

In [189]:
df.describe()

,Price
count,29773.000000
mean,23.392503
std,29.949671
min,1.000000
25%,5.000000
50%,11.000000
75%,35.000000
max,267.000000


<hr>

Выполним срез и сгруппируем по методу платежа, после чего вызовем агрегатную функцию count, чтобы подсчитать количеству транзакций по тому или иному методу.

Исходя из вывода можно сделать, что большинство покупают через кредитную карточку.

In [190]:
df_group = df[['Transaction ID', 'Payment Method']]
df_group.groupby('Payment Method').count()

,Transaction ID
Payment Method,
Contactless,10284
Credit Card,17881
Debit Card,1608


<hr>

Выполним срез и оставим в нем только те случаи в которых отказались от билетов. После чего посчитаем с каких станций больше всего люди отменяли билеты.

Исходя из результата топ-5 больше всего отказов в `Liverpool Lime Street`.

In [191]:
df_group = df[['Departure Station', 'Refund Request', 'Journey Status']]

cancelled_tickets = df_group[df_group['Refund Request'].str.contains(
    'Yes', case=False)]

cancelled_tickets['Departure Station'].value_counts()[0:5]

Departure Station
Liverpool Lime Street    232
Birmingham New Street    109
Manchester Piccadilly    108
Edinburgh Waverley        51
London Euston             17
Name: count, dtype: int64

Посмотрим какие уникальные значения содержит столбец `Journey Status`.

In [192]:
df['Journey Status'].unique()

array(['On Time', 'Delayed'], dtype=object)

Посмотрим сколько уникальных значений у `Journey Status` после группировки. По результатам оказывается что только одно значение `Delayed`. А это значит что во всех случаях, когда возвращали билет поезд опаздывал.

In [193]:
cancelled_tickets['Journey Status'].unique()

array(['Delayed'], dtype=object)

Выполним срез и оставим в нем только те случаи в которых поезда опоздали. После чего посчитаем на каких станциях поезда чаще всего опаздывают.

Исходя из результата топ-5 больше всего опозданий в `Liverpool Lime Street`.

In [194]:
df_group = df[['Departure Station', 'Refund Request', 'Journey Status']]

delayed_train = df_group[df_group['Journey Status'].str.contains(
    'Delayed', case=False)]

delayed_train['Departure Station'].value_counts()[0:5]

Departure Station
Liverpool Lime Street    900
Manchester Piccadilly    672
London Euston            259
Birmingham New Street    140
London Kings Cross       131
Name: count, dtype: int64

Вывод: Был проведен разведочный анализ данных по датасету. Было изучено как группировать, фильтровать, сортировать, обрабатывать, делать срезы данных.